keps/sig-storage/2644-honor-pv-reclaim-policy/README.md

```diff
@@ -0,0 +1,246 @@
+	// read current volume state now.
+	newVolume, err := ctrl.kubeClient.CoreV1().PersistentVolumes().Get(context.TODO(), volume.Name, metav1.GetOptions{})
+	if err != nil {
+		klog.V(3).Infof("error reading persistent volume %q: %v", volume.Name, err)
+		return "", nil
+	}
+
+	if newVolume.GetDeletionTimestamp() != nil {//<==========================Remove check.
+		klog.V(3).Infof("Volume %q is already being deleted", volume.Name)
+		return "", nil
+	}
+	needsReclaim, err := ctrl.isVolumeReleased(newVolume)
+	if err != nil {
+		klog.V(3).Infof("error reading claim for volume %q: %v", volume.Name, err)
+		return "", nil
+	}
+	if !needsReclaim {
+		klog.V(3).Infof("volume %q no longer needs deletion, skipping", volume.Name)
+		return "", nil
+	}
+
+	pluginName, deleted, err := ctrl.doDeleteVolume(volume)
+	// Remaining code below skipped for brevity...
+	// Please refer pkg/controller/volume/persistentvolume/pv_controller.go in kubernetes/kubernetes for the full code. 
+		
+}
+```
+
+The proposed fix will not have side effects in existing code flows as the `deleteVolumeOperation` is scheduled with an
+identifier generated using the volume name and UID that is being deleted. All duplicate identifiers are skipped.
+
+```go
+	case v1.PersistentVolumeReclaimDelete:
+		klog.V(4).Infof("reclaimVolume[%s]: policy is Delete", volume.Name)
+		opName := fmt.Sprintf("delete-%s[%s]", volume.Name, string(volume.UID))
+		// create a start timestamp entry in cache for deletion operation if no one exists with
+		// key = volume.Name, pluginName = provisionerName, operation = "delete"
+		ctrl.operationTimestamps.AddIfNotExist(volume.Name, ctrl.getProvisionerNameFromVolume(volume), "delete")
+		ctrl.scheduleOperation(opName, func() error {
+			_, err := ctrl.deleteVolumeOperation(volume)
+			if err != nil {
+				// only report error count to "volume_operation_total_errors"
+				// latency reporting will happen when the volume get finally
+				// deleted and a volume deleted event is captured
+				metrics.RecordMetric(volume.Name, &ctrl.operationTimestamps, err)
+			}
+			return err
+		})
+```
+
+### Test Plan
+
+#### E2E tests
+An e2e test to exercise deletion of PV prior to deletion of PVC for a `Bound` PV-PVC pair.
+
+
+### Graduation Criteria
+
+#### Alpha -> Beta
+* Feedback
+* Fix of the issue behind a feature flag.
+* Unit tests and e2e tests outlined in design proposal implemented.
+
+#### Beta -> GA
+* Deployed in production and have gone through at least one K8s upgrade.
+
+### Upgrade / Downgrade Strategy
+* Upgrade from old Kubernetes(1.21) to new Kubernetes(1.22) with `HonorPvReclaim` flag enabled
+
+In the above scenario, the upgrade will cause change in default behavior as described in the current KEP. Additionally,
+no efforts are made to delete volumes that were not previously as a result of old behavior.
+
+* Downgrade from new Kubernetes(1.22) to old Kubernetes(1.21) with `HonorPvReclaim` flag disabled
```

Comment:
Does this imply you would disable the flag first, then downgrade?

Reply:
I don't fully understand how flags work here, my assumption is once downgraded from 1.22 to 1.21, the flag `HonorPvReclaim` becomes irrelevant, I am assuming that downgrading after disabling or downgrading with it enabled does not make a difference. I updated it to reflect this. I may be wrong, hopefully, you can correct me on this.